src/main/java/org/commcare/session/CommCareSession.java

```diff
@@ -116,74 +117,92 @@ public CommCareSession(CommCareSession oldCommCareSession) {
      * @return A list of all of the form entry actions that are possible with the given commandId
      * and the given list of already-collected datums
      */
     private Vector<Entry> getEntriesForCommand(String commandId,
-                                               OrderedHashtable<String, String> currentSessionData) {
+                                               OrderedHashtable<String, String> currentSessionData,
+                                               boolean includeNested) {
         Vector<Entry> entries = new Vector<>();
         if (commandId == null) {
             return entries;
         }
         for (Suite s : platform.getInstalledSuites()) {
             List<Menu> menusWithId = s.getMenusWithId(commandId);
             if (menusWithId != null) {
-                for (Menu menu : menusWithId) {
-                    entries.addAll(getEntriesFromMenu(menu, currentSessionData));
+                // make a copy so that we can use this as a queue
+                List<Menu> menusToExamine = new ArrayList<>(menusWithId);
+                while (!menusToExamine.isEmpty()) {
+                    Menu menu = menusToExamine.remove(0);
+                    entries.addAll(getStillValidEntriesFromMenu(menu, currentSessionData));
+                    if (includeNested) {
+                        menusToExamine.addAll(s.getMenusWithRoot(menu.getId()));
+                    }
                 }
             }
 
             if (s.getEntries().containsKey(commandId)) {
                 entries.addElement(s.getEntries().get(commandId));
             }
         }
-
         return entries;
     }
 
-    /**
-     * Get all entries that correspond to commands listed in the menu provided.
-     * Excludes entries whose data requirements aren't met by the 'currentSessionData'
-     */
-    private Vector<Entry> getEntriesFromMenu(Menu menu,
-                                             OrderedHashtable<String, String> currentSessionData) {
-        Vector<Entry> entries = new Vector<>();
-        Hashtable<String, Entry> map = platform.getMenuMap();
-        //We're in a menu we have a set of requirements which
-        //need to be fulfilled
+    private Vector<Entry> getStillValidEntriesFromMenu(Menu menu,
+                                                       OrderedHashtable<String, String> currentSessionData) {
+        Hashtable<String, Entry> globalEntryMap = platform.getCommandToEntryMap();
+        Vector<Entry> stillValid = new Vector<>();
         for (String cmd : menu.getCommandIds()) {
-            Entry e = map.get(cmd);
+            Entry e = globalEntryMap.get(cmd);
             if (e == null) {
                 throw new RuntimeException("No entry found for menu command [" + cmd + "]");
             }
-            if (entryRequirementsSatsified(e, currentSessionData)) {
-                entries.addElement(e);
+            if (entryRequiresAllDataInSession(e, currentSessionData)) {
```

Comment:
The more I think about this, the less it makes sense to me.

Specifically in the end of end-of-form navigation or register from case list (entity list <action>'s) you could add the session frame
```
m_0
case_id_one
case_id_two
```
to the session, and have two menu declarations you are jumping to, each with an entry:
```
m_0
e_1
 - case_id_one
 - case_id_two

m_0
e_2
 - case_id_two
 - case_id_one
```
and e_1 and e_2 should both meet this check. If you visited the `m_0` form directly, you'd expect to need to choose e_1 or e_2 first, because their order of selecting id's is different (and they may have different filters), but once you've chosen both id's, semantically both forms should accept the two.

Reply:
Hey sorry, I'm not 100% sure I see the conclusion you're leading to with that, in terms of what you think is wrong with the current implementation. Are you objecting to the fact that `entryRequiresAllDataInSession()` insists that the matching datum IDs match _in order_? If so, I think I agree that that would be a problem in the edge case you gave. But is that all you meant, or is there something else/more than that?

Reply:
And just to make note of it now- we'd be taking a small performance hit if we stopped doing the in-order check and just did N `contains` checks instead. Given that N is never large in the context of # of datums on the session stack, I think this is totally fine, but I had been assuming that that was part of why it was done that way to begin with.

Reply:
Hey,

Yeah, the in-order check is the most relevant part. I'm also honestly not clear on whether

```
case_id_one
case_id_three
case_id_two
```

_shouldn't_ match something that needs
```
case_id_one
case_id_two
```

Based on how we're using this code, which is to decide whether an entry is relevant, right? I think the only thing that makes an entry irrelevant based on what happens in the session are commands, rather than datums.

Reply:
Hm so I think I'm coming to a similar conclusion, but with slightly different reasoning (or maybe we are saying the same thing in different ways). But my thinking is this:

One of the contracts we adhere to in the CommCareSession is that we never have a user select a datum (or have the engine perform the computation of a computed datum) that isn't going to be used/needed by whatever form the user ends up at in that session. This is guaranteed by [this line](https://github.com/dimagi/commcare-core/blob/master/src/main/java/org/commcare/session/CommCareSession.java#L207), where we only return a datum from `getNeededData()` if it's required by _all_ the remaining valid entries. So on the assumption that there are no flaws in that implementation, then I think we can just say that the check performed by `entryRequiresAllDataInSession()` is repetitive, and therefore not needed. Does that reasoning sound right to you?